### PR TITLE
Use DNSRecord resources

### PR DESCRIPTION
**Pull request by stoyanr**
**How to categorize this PR?**

/area control-plane
/kind enhancement

**What this PR does / why we need it**:
* Introduces a component and helper functions for handling `DNSRecord` resources, previously introduced with #4231.
* Changes the shoot reconcile, delete, and migrate flows to use the new `DNSRecord` resource instead of the old `DNSProvider`, `DNSOwner`, and `DNSEntry` resources.
* Adds support for installing the correct DNSRecord extension in the seed.
* Adds support for explicit zones via an annotation on domain secrets.
* Changes the seed reconcile and delete flow to use the new `DNSRecord` resource as well if the managed ingress is enabled for the seed (including support for installing the correct DNSRecord extension).
* Adds a feature gate and support for migration of DNS resources for shoots and for seeds.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

* Based on #4231, marked as Draft until that one is merged.
* The handling of owner TXT records that was previously added to this PR is now removed. It will be added in a new PR.
* No unit tests or documentation yet, I will add these as separate commits.

**Release note**:

```other operator
NONE
```

### CODE REVIEW

File: pkg/utils/gardener/dns.go
Comment (MartinWeindel) [inline pkg/utils/gardener/dns.go:54]: 6-Tuple as return value is error-prone. Wouldn't it be clearer to return a struct with the fields provider,domain,zone,includeZones,excludeZones ?

Comment (stoyanr) [reply pkg/utils/gardener/dns.go:54]: Well, I agree, but I am adapting existing code. 6-tuple is not more error-prone than 5-tuple, and this 5-tuple is used in a dozen places at least. I would keep it as it is in this PR and perhaps refactor it in a separate PR.

File: pkg/operation/botanist/addons.go
Comment (timebertt) [inline pkg/operation/botanist/addons.go]: `common.ShootDNSIngressName` should be enough, the name is already contained in the namespace.

Pros:
- shorter names
- same name for resources for all namespaces, makes it easier for operators as commands can be reused across namespaces (e.g. from history / in loops / with xargs / ...)

Cons:
- naming scheme is consistent with other extension resource names (e.g. `ControlPlane` resources, one is named `<shoot-name>` the other one is named `<shoot-name>-exposure`)

@rfranzke for more opinions :)

File: pkg/operation/botanist/dnsrecord.go
Comment (timebertt) [inline pkg/operation/botanist/dnsrecord.go]: `DNSExternalName` should be enough, the name is already contained in the namespace.

File: pkg/operation/botanist/dnsrecord.go
Comment (timebertt) [inline pkg/operation/botanist/dnsrecord.go]: `DNSInternalName` should be enough, the name is already contained in the namespace.

File: pkg/operation/botanist/addons.go
Comment (timebertt) [inline pkg/operation/botanist/addons.go]:
```suggestion
// DefaultIngressDNSRecord creates the default deployer for the ingress DNSRecord resource.
```

File: pkg/operation/seed/seed.go
Comment (timebertt) [inline pkg/operation/seed/seed.go]: Please put some `//TODO`s here to say that this is a hack and we want to make the contract between gardenlet and extension controllers for reconciling resources related to the seed more explicit in the future.

File: pkg/operation/seed/seed.go
Comment (timebertt) [inline pkg/operation/seed/seed.go]: Use `controllerutils.GetAndCreateOrMergePatch` instead, or simply rebase onto the latest master.

File: pkg/features/features.go
Comment (timebertt) [inline pkg/features/features.go]: Will most likely not be in `v1.26`
```suggestion
	// alpha: v1.27.0
```

File: pkg/features/features.go
Comment (timebertt) [inline pkg/features/features.go:99]: Add to list of feature gates in https://github.com/gardener/gardener/blob/master/docs/deployment/feature_gates.md, link detailed documentation.

File: pkg/controllermanager/controller/controllerregistration/controllerregistration_seed_control_test.go
Comment (timebertt) [inline pkg/controllermanager/controller/controllerregistration/controllerregistration_seed_control_test.go]: Let's disable the feature gate in the tests (which is the default currently), and flip it on once we enabled it by default.

File: pkg/operation/botanist/dns_test.go
Comment (timebertt) [inline pkg/operation/botanist/dns_test.go]: Let's disable the feature gate in the tests (which is the default currently), and flip it on once we enabled it by default.

File: pkg/operation/botanist/dnsresources.go
Comment (timebertt) [inline pkg/operation/botanist/dnsresources.go:32]: Doc string / comment / explanation?

File: pkg/operation/shoot/shoot_test.go
Comment (timebertt) [inline pkg/operation/shoot/shoot_test.go]: Test with false (same like the feature gate comment above)

File: pkg/controllerutils/finalizers.go
Comment (timebertt) [inline pkg/controllerutils/finalizers.go:61]: Tests?

File: pkg/operation/seed/seed.go
Comment (timebertt) [inline pkg/operation/seed/seed.go:903]: No reason to get secret again, is already read in `updateDNSProviderSecret`.
Let's get the `secretData` first and then pass it down to `updateDNSProviderSecret` (which is only needed if the `UseDNSRecords` feature gate is disabled).

File: pkg/operation/seed/seed.go
Comment (timebertt) [inline pkg/operation/seed/seed.go:1022]: No reason to get secret again, is already read in `updateDNSProviderSecret`.
Let's get the `secretData` first and then pass it down to `updateDNSProviderSecret` (which is only needed if the `UseDNSRecords` feature gate is disabled).

File: pkg/operation/seed/seed.go
Comment (timebertt) [inline pkg/operation/seed/seed.go]: Use `controllerutils.GetAndCreateOrMergePatch` instead.

File: pkg/operation/shoot/shoot.go
Comment (timebertt) [inline pkg/operation/shoot/shoot.go]: We need similar adaptions to the `ExtensionValidator` admission plugin:
- shoot dns provider cross-check (if there is a primary provider we additionally need the `DNSRecord`, `DNSProvider` is still required for all providers specified in `spec.dns.providers`): https://github.com/gardener/gardener/blob/e825f740eb93f7f9497bac0a31b01e8cc2dbb019/plugin/pkg/global/extensionvalidation/admission.go#L246
- seed dns provider cross-check for managed ingress DNS (is completely missing): https://github.com/gardener/gardener/blob/e825f740eb93f7f9497bac0a31b01e8cc2dbb019/plugin/pkg/global/extensionvalidation/admission.go#L206

**Review by timebertt — COMMENTED** [review]
Here is the feedback, that we just discussed. :)
For others reading up on this: there are mainly minor things right now about the current state of the PR (see my inline comments), rest looks very good already.

File: pkg/operation/botanist/addons.go
Comment (stoyanr) [reply pkg/operation/botanist/addons.go]: I personally like consistent naming more than brevity and would keep the names as they are currently. If however there are more votes for the shorter names I will of course rename as proposed.

File: pkg/operation/seed/seed.go
Comment (stoyanr) [reply pkg/operation/seed/seed.go]: Done (rebased on the latest master).

File: pkg/features/features.go
Comment (stoyanr) [reply pkg/features/features.go]: Changed to v.1.27.0 as proposed.

File: pkg/controllermanager/controller/controllerregistration/controllerregistration_seed_control_test.go
Comment (stoyanr) [reply pkg/controllermanager/controller/controllerregistration/controllerregistration_seed_control_test.go]: Done.

File: pkg/operation/shoot/shoot_test.go
Comment (stoyanr) [reply pkg/operation/shoot/shoot_test.go]: Done.

File: pkg/operation/seed/seed.go
Comment (stoyanr) [reply pkg/operation/seed/seed.go]: Done (rebased on the latest master).

File: pkg/operation/botanist/dns_test.go
Comment (stoyanr) [reply pkg/operation/botanist/dns_test.go]: Done.

File: pkg/operation/botanist/addons.go
Comment (stoyanr) [reply pkg/operation/botanist/addons.go]: Done.

File: pkg/operation/seed/seed.go
Comment (stoyanr) [reply pkg/operation/seed/seed.go]: Done.

File: pkg/features/features.go
Comment (stoyanr) [reply pkg/features/features.go:99]: Done.

File: pkg/operation/botanist/dnsresources.go
Comment (stoyanr) [reply pkg/operation/botanist/dnsresources.go:32]: Add a detailed doc string for this and all other methods in this file.

File: pkg/operation/seed/seed.go
Comment (stoyanr) [reply pkg/operation/seed/seed.go:903]: Done.

File: pkg/operation/seed/seed.go
Comment (stoyanr) [reply pkg/operation/seed/seed.go:1022]: Done.

File: pkg/operation/shoot/shoot.go
Comment (stoyanr) [reply pkg/operation/shoot/shoot.go]: Done.

**Comment by stoyanr** [comment]
Restructured and consolidated the commits of this PR in order to avoid removing code and reintroducing it later, to consolidate related commits into one, and to remove the commit that introduced the owner record. Also added a new commit to address most of the code review comments.
@timebertt I believe I addressed all your feedback except the naming of the DNSRecord resources (I am waiting for more input) and the missing tests (I am going to introduce them and all other missing unit tests in a new commit).

File: pkg/controllerutils/finalizers.go
Comment (stoyanr) [reply pkg/controllerutils/finalizers.go:61]: Done.

**Comment by stoyanr** [comment]
Added a new commit with the needed unit tests. I consider this PR ready now except for the fact that it's based on #4231, which is still not merged, so keeping the Draft state until this is the case.

**Comment by timebertt** [comment]
#4231 is merged. This one can be rebased.

**Comment by stoyanr** [comment]
> #4231 is merged. This one can be rebased.

Thanks! Rebased it and removed `Draft` now.

File: pkg/operation/botanist/component/extensions/dnsrecord/dnsrecord_test.go
Comment (timebertt) [inline pkg/operation/botanist/component/extensions/dnsrecord/dnsrecord_test.go]: Nit: generally it's nicer to use a fake implementation of `retry.Ops` rather than relying on short intervals/timeouts (makes tests faster, more robust in CI environments). E.g. like:
https://github.com/gardener/gardener/blob/712a875b5965c9c87a8ce9c304c13733fe4b11ce/pkg/operation/botanist/component/extensions/dns/dnsentry_test.go#L72

File: docs/deployment/feature_gates.md
Comment (timebertt) [inline docs/deployment/feature_gates.md:93]: Can you add some explanations to `docs/extensions/dnsrecord.md` about what will happen when you turn on the `UseDNSRecords` feature gate, what is the default behavior (i.e. when it's disabled) and what is required to turn it on (e.g. `ControllerRegistration` for `DNSRecord`s for default, internal and all custom domain providers required) and so on?

**Review by timebertt — COMMENTED** [review]
Nice! Thanks for incorporating my suggestions :)
Not much left to add from my side, codewise it looks very good.

I'm now heading over to another round of review of https://github.com/gardener/gardener-extension-provider-aws/pull/362 and then I want to test both PRs together in my local setup and play around with it a bit.

**Comment by stoyanr** [comment]
> I'm now heading over to another round of review of gardener/gardener-extension-provider-aws#362 and then I want to test both PRs together in my local setup and play around with it a bit.

Before you start testing, perhaps you could wait until #4312 is merged and I have vendored it in gardener/gardener-extension-provider-aws#362. My fork contains mostly the same code, but just in case.

**Comment by timebertt** [comment]
@gardener/gardener-maintainers Can you add some more opinions on the names of the `DNSRecord`s? https://github.com/gardener/gardener/pull/4253#discussion_r660600216
I think, this is the only open point left from my previous review.

File: pkg/operation/botanist/dnsresources_test.go
Comment (MartinWeindel) [inline pkg/operation/botanist/dnsresources_test.go]: The external `DNSProvider` must not be deleted as it may still be needed for `DNSEntries` of the `shoot-dns-service`. The internal `DNSProvider` can be deleted, but not the external one.

**Review by MartinWeindel — CHANGES_REQUESTED** [review]
The external DNSProvider must no be deleted.

File: docs/deployment/feature_gates.md
Comment (stoyanr) [reply docs/deployment/feature_gates.md:93]: Done

File: pkg/operation/botanist/component/extensions/dnsrecord/dnsrecord_test.go
Comment (stoyanr) [reply pkg/operation/botanist/component/extensions/dnsrecord/dnsrecord_test.go]: Done

File: pkg/operation/botanist/dnsresources_test.go
Comment (MartinWeindel) [inline pkg/operation/botanist/dnsresources_test.go:190]: The `DNSOwner` should be created after the `DNSEntry`. With the current order it can happen (and happened on my test env), that the DNS record on the DNS server is deleted by the `dns-external` extension first, before it is recreated as soon as the `DNSEntry` has be created. This is only relevant, if the owner TXT DNS record is still present when switching back from `UseDNSRecords=true` to `UseDNSRecords=false`.
As a consequence, the Kube-apiserver may be unavailable via both internal and external domain name for several minutes depending on the negative DNS caching on the client side.

Comment (stoyanr) [reply pkg/operation/botanist/dnsresources_test.go:190]: This order comes from the existing implementation in `DeployExternalDNS`. I would rather not do something special here, since what I want to do is "deploy the external DNS resources", so I can't understand how the existing implementation can be wrong if it worked so far? Or is it really only wrong because the the owner TXT DNS record is still present? In this case, if we take care to delete this record when switching to `UseDNSRecords=true`, so that it's not present in the case you mentioned, everything would be ok, correct?

These owner TXT records should be deleted anyway. I plan to add the needed fix to the AWS provider extensions (and all other provider extensions). I intentionally don't want to do something special for them in the Gardener code and leave that to the extensions, the picture in Gardener is quite complicated and hard to understand already.

File: pkg/operation/botanist/dnsresources_test.go
Comment (stoyanr) [reply pkg/operation/botanist/dnsresources_test.go]: If I understand this correctly, if it can be used for `DNSEntries` of the `shoot-dns-service`, not only should it not be deleted, but it should be also **created** if a new cluster is created with `UseDNSRecords=true`, correct? So far I thought that such a `DNSProvider` should be created for the "primary" provider listed in `spec.dns.providers`, if there is such a provider, but not if there is no such provider and a "default domain" secret is used. Should it really also be created for the "default domain" secret case as well?

@timebertt @rfranzke Could you please help here? I am not sure I fully understand what exactly is supported right now. Let's clarify this before I do further changes.

Comment (stoyanr) [reply pkg/operation/botanist/dnsresources_test.go]: My thoughts on the matter:

* If the above assumption is true, then the external `DNSProvider` should always be created (and not touched when migrating from `UseDNSRecords=true` to `UseDNSRecords=false` and vice versa), either from the "primary" provider in the shoot spec or from the "default domain" secret. This however is contradictory to the idea to completely decouple Gardener from `DNSProviders` in the future, so we should think carefully if we really need this.
* If the above assumption is not true, then the  external `DNSProvider` must only be created if there is a "primary" provider in the shoot spec. In the migration case, the above logic may delete it, but it will still be created as an "additional provider" later and will therefore still exist after the migration. I assume this should be ok in this case?

Comment (timebertt) [reply pkg/operation/botanist/dnsresources_test.go]: I think, you understood correctly, @stoyanr.
However, @MartinWeindel has a good point here (nice catch, didn't think about this case in my reviews): currently, the default dns provider can be used by the shoot-dns-service for DNSEntries in the shoot as well, because of the implicit dependencies involved here.
So we probably indeed need to keep/reconcile the `DNSProvider` for the default domain as well for backwards-compatibility.

Comment (rfranzke) [reply pkg/operation/botanist/dnsresources_test.go]: Hm, I couldn't follow this in detail from the comments, but why do we have to keep this logic in gardenlet? Can't we rather change the `shoot-dns-service` extension to take over responsibility for the "external" `DNSProvider`? In the migration case (`UseDNSRecords=true`) we could simply not delete it?

File: pkg/operation/botanist/dnsresources_test.go
Comment (MartinWeindel) [reply pkg/operation/botanist/dnsresources_test.go:190]: Correct, if you take care to delete the owner TXT records when switching to `UseDNSRecords=true`, everything should be ok.

File: pkg/operation/botanist/dnsresources_test.go
Comment (MartinWeindel) [reply pkg/operation/botanist/dnsresources_test.go]: This is my understanding about shifting the responsibility for the `DNSProvider` to the `shoot-dns-service` after a private chat with @stoyanr:
- it should be dealt with only after the dust of the migration to `DNSRecord` has settled 
- the specification of the DNS providers should be migrated from the `spec.dns.providers` section in the shoot manifest to the `spec.extensions[type=shoot-dns-service].providerConfig.providers` section
- the provider secrets located in the virtual garden must be made available to the extension, e.g. by copying them to the CP using "referenced resources"

Comment (timebertt) [reply pkg/operation/botanist/dnsresources_test.go]: > This is my understanding about shifting the responsibility for the `DNSProvider` to the `shoot-dns-service` after a private chat with @stoyanr:

Yes, I agree, this is also what I had in mind. We will need a second migration step for the `DNSProvider`s in order to resolve this implicit coupling between gardener and shoot-dns-service.
But let's focus on the `DNSRecord`s only for now, not touch the `DNSProvider`s and take care about the shift for the `DNSProvider`s later on 👍

Comment (stoyanr) [reply pkg/operation/botanist/dnsresources_test.go]: As agreed, I added logic to ensure that the external DNSProvider is always created (also when `UseDNSRecords=true`), and removed the special handling of the primary DNS provider in `AdditionalDNSProviders` that was previously introduced. Still testing all possible cases.

File: pkg/operation/botanist/dnsresources_test.go
Comment (stoyanr) [reply pkg/operation/botanist/dnsresources_test.go:190]: OK, then we'll take care to delete them in the corresponding provider extensions.

File: pkg/operation/botanist/dnsresources_test.go
Comment (stoyanr) [reply pkg/operation/botanist/dnsresources_test.go]: I tested the following cases and it seems to work ok:
* No DNS providers in shoot spec - the external `DNSProvider` is still created from the default-domain secret
* One non-primary DNS provider - both the external `DNSProvider` and the non-primary `DNSProvider` are created.
* One primary DNS provider - only the primary `DNSProvider` is created from the info in the shoot spec.

I also tested hibernation, CP migration, etc. for the third case.

**Comment by stoyanr** [comment]
As discussed, added the `DNSOwner` for the seed managed ingress `DNSEntry` in order to be able to delete this `DNSEntry` without actually removing the DNS record. Now this PR is really ready for further testing / merge. 
/cc @timebertt

**Review by timebertt — DISMISSED** [review]
Thank you very much!
Code-wise 
/lgtm
now!

Won't been able to test it myself e2e, but you have already completed extensive testing.

File: pkg/operation/botanist/addons.go
Comment (rfranzke) [inline pkg/operation/botanist/addons.go:138]: Nit: doc string missing

File: pkg/operation/botanist/component/extensions/dnsrecord/dnsrecord_suite_test.go
Comment (rfranzke) [inline pkg/operation/botanist/component/extensions/dnsrecord/dnsrecord_suite_test.go]:
```suggestion
func TestDNSRecord(t *testing.T) {
```

**Review by rfranzke — DISMISSED** [review]
I have only had a rather superficial look, but what I saw
/lgtm

Thanks a lot!

File: pkg/operation/botanist/addons.go
Comment (stoyanr) [reply pkg/operation/botanist/addons.go:138]: Done

File: pkg/operation/botanist/component/extensions/dnsrecord/dnsrecord_suite_test.go
Comment (stoyanr) [reply pkg/operation/botanist/component/extensions/dnsrecord/dnsrecord_suite_test.go]: Done

**Review by timebertt — DISMISSED** [review]
/lgtm
/milestone v1.27

**Review by timebertt — APPROVED** [review]
Well, there's simply no way to review the changes from the rebase. So I'll have to trust you on this one :)
/lgtm

**Comment by timebertt** [comment]
@MartinWeindel I would like to include this in v1.27. 
Do you plan to take another look?

**Comment by timebertt** [comment]
Oh, I see @MartinWeindel is on vacation this week.
So merging this now to include it in the final release validation for v1.27.

**Pull request merged**
